#Check asteroid data

In [1]:
import pandas as pd

In [8]:
jpl_df = pd.read_csv("Data/JPL_results_20170408.csv", low_memory=False, index_col='id')

print("rows =%d"%len(jpl_df))
print("columns = %d"%len(jpl_df.columns))

rows =732354
columns = 74


In [48]:
jpl_df[['pdes','full_name']].loc['a0138911']

pdes                    138911
full_name    138911 (2001 AE2)
Name: a0138911, dtype: object

In [49]:
jpl_df[['pdes','full_name']].head(3)

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pdes</th>
      <th>full_name</th>
    </tr>
    <tr>
      <th>id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>a0000001</th>
      <td>1</td>
      <td>1 Ceres</td>
    </tr>
    <tr>
      <th>a0000002</th>
      <td>2</td>
      <td>2 Pallas</td>
    </tr>
    <tr>
      <th>a0000003</th>
      <td>3</td>
      <td>3 Juno</td>
    </tr>
  </tbody>
</table>
</div>

In [57]:
len(jpl_df[['pdes','full_name']][jpl_df['full_name'].str.contains('\(')])

732022

It looks like full_name is a combination of pdes and prov_des

In [10]:
asteroid_rank_columns = {
      "sigma_tp":4.4133e-05,
      "diameter":"",
      "tp":2456656.02501109,
      "epoch_mjd":56400.0,
      "ad":1.459936639399153,
      "producer":"Otto Matic",
      "rms":0.57773,
      "H_sigma":0.66439,
      "closeness":4.659571917281135,
      "spec":"T",
      "K2":"",
      "K1":"",
      "M1":"",
      "n_dop_obs_used":"",
      "full_name":"138911 (2001 AE2)",
      "M2":"",
      "prov_des":"2001 AE2",
      "equinox":"J2000",
      "DT":"",
      "diameter_sigma":"",
      "saved":-8641518801655.289,
      "albedo":"",
      "moid_ld":90.96226078,
      "two_body":"",
      "neo":"Y",
      "prefix":"",
      "sigma_ad":3.3626e-09,
      "score":3.4508560938390502e-06,
      "data_arc":10120.0,
      "profit":2146841.445758379,
      "sigma_q":5.7701e-08,
      "sigma_w":0.00015086,
      "epoch":2456400.5,
      "per":572.7291693097955,
      "id":"a0138911",
      "A1":"",
      "PC":"",
      "A3":"",
      "A2":"",
      "per_y":1.5680470070083,
      "n_opp":"",
      "epoch_cal":20130418.0,
      "orbit_id":"JPL 82",
      "sigma_a":3.1086e-09,
      "sigma_om":0.00014823,
      "price":17254280.46919525,
      "sigma_e":4.2218e-08,
      "condition_code":0.0,
      "rot_per":"",
      "G":"",
      "last_obs":"2012-04-17",
      "H":19.032,
      "pha":"N",
      "IR":"",
      "inexact":True,
      "spec_T":"",
      "tp_cal":20131229.5250111,
      "n_obs_used":381.0,
      "moid":0.233734,
      "extent":"",
      "dv":4.057,
      "ma":199.3848106194188,
      "GM":"",
      "pdes":138911.0,
      "class":"AMO",
      "a":1.349689648496459,
      "t_jup":4.87,
      "om":171.4867878184716,
      "e":0.08168321586040721,
      "name":"",
      "i":1.661150372152747,
      "sigma_per":1.9787e-06,
      "BV":"",
      "n":0.6285693470682512,
      "q":1.239442657593766,
      "sigma_i":3.6361e-06,
      "w":43.09771100599832,
      "sigma_ma":2.7756e-05,
      "first_obs":"1984-08-02",
      "n_del_obs_used":"",
      "sigma_n":2.1716e-09,
      "spkid":2138911.0,
      "UB":""
   }

In [11]:
for col in jpl_df.columns:
    if col not in asteroid_rank_columns.keys():
        print(col)

spec_B
moid_jup


it looks like we have some extra columns in jpl

In [12]:
for col in asteroid_rank_columns:
    if col not in jpl_df.columns:
        print(col)

closeness
spec
prov_des
saved
score
profit
id
n_opp
price
inexact
dv


AsteroidRank has some extra stuff.

We can get dv from their github.

score and price seem to be their predict values. We shall investigate the others later.

In [21]:
col_count = jpl_df.count()
zero_cols = []
for key, value in dict(col_count).items():
    if value==0:
        print(key)
        zero_cols.append(key)
        if key in asteroid_rank_columns:
            print("\t but in asteroind_rank_columns")

prefix
	 but in asteroind_rank_columns
M1
	 but in asteroind_rank_columns
M2
	 but in asteroind_rank_columns
K1
	 but in asteroind_rank_columns
K2
	 but in asteroind_rank_columns
PC
	 but in asteroind_rank_columns
A3
	 but in asteroind_rank_columns
DT
	 but in asteroind_rank_columns


These look like Comet features, probably in AsteroidRank for API simplification

In [23]:
jpl_df2 = jpl_df.drop(zero_cols,axis=1)
jpl_df2.columns

Index(['spkid', 'full_name', 'pdes', 'name', 'neo', 'pha', 'H', 'G',
       'diameter', 'extent', 'albedo', 'rot_per', 'GM', 'BV', 'UB', 'IR',
       'spec_B', 'spec_T', 'H_sigma', 'diameter_sigma', 'orbit_id', 'epoch',
       'epoch_mjd', 'epoch_cal', 'equinox', 'e', 'a', 'q', 'i', 'om', 'w',
       'ma', 'ad', 'n', 'tp', 'tp_cal', 'per', 'per_y', 'moid', 'moid_ld',
       'moid_jup', 't_jup', 'sigma_e', 'sigma_a', 'sigma_q', 'sigma_i',
       'sigma_om', 'sigma_w', 'sigma_ma', 'sigma_ad', 'sigma_n', 'sigma_tp',
       'sigma_per', 'class', 'producer', 'data_arc', 'first_obs', 'last_obs',
       'n_obs_used', 'n_del_obs_used', 'n_dop_obs_used', 'condition_code',
       'rms', 'two_body', 'A1', 'A2'],
      dtype='object')

In [73]:
print("rows =%d"%len(jpl_df2))
print("columns = %d"%len(jpl_df2.columns))

rows =732354
columns = 67


In [26]:
asterank_dv_df = pd.read_csv("Data/Asterank_computed_dv_20170408.csv",low_memory=False,index_col='pdes')

print("rows =%d"%len(asterank_dv_df))
print("columns = %d"%len(asterank_dv_df.columns))

rows =730407
columns = 1


In [93]:
asterank_dv_df.sort_index(ascending=True).head(100)

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dv</th>
    </tr>
    <tr>
      <th>pdes</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>10.533885</td>
    </tr>
    <tr>
      <th>10</th>
      <td>10.461966</td>
    </tr>
    <tr>
      <th>100</th>
      <td>10.241299</td>
    </tr>
    <tr>
      <th>1000</th>
      <td>11.456043</td>
    </tr>
    <tr>
      <th>10000</th>
      <td>10.387258</td>
    </tr>
    <tr>
      <th>100000</th>
      <td>10.534720</td>
    </tr>
    <tr>
      <th>100001</th>
      <td>9.489282</td>
    </tr>
    <tr>
      <th>100002</th>
      <td>8.693603</td>
    </tr>
    <tr>
      <th>100003</th>
      <td>8.420975</td>
    </tr>
    <tr>
      <th>100004</th>
      <td>8.658539</td>
    </tr>
    <tr>
      <th>100005</th>
      <td>8.634136</td>
    </tr>
    <tr>
      <th>100006</th>
      <td>7.180624</td>
    </tr>
    <tr>
      <th>100007</th>
      <td>12.931505</td>
    </tr>
    <tr>
      <th>100008</th>
      <td>8.606221</td>
    </tr>
    <tr>
      <th>100009</th>
      <td>8.157409</td>
    </tr>
    <tr>
      <th>10001</th>
      <td>9.215672</td>
    </tr>
    <tr>
      <th>100010</th>
      <td>10.297042</td>
    </tr>
    <tr>
      <th>100011</th>
      <td>10.123094</td>
    </tr>
    <tr>
      <th>100012</th>
      <td>12.087700</td>
    </tr>
    <tr>
      <th>100013</th>
      <td>10.247814</td>
    </tr>
    <tr>
      <th>100014</th>
      <td>9.141752</td>
    </tr>
    <tr>
      <th>100015</th>
      <td>8.114382</td>
    </tr>
    <tr>
      <th>100016</th>
      <td>8.721262</td>
    </tr>
    <tr>
      <th>100017</th>
      <td>8.212097</td>
    </tr>
    <tr>
      <th>100018</th>
      <td>8.886286</td>
    </tr>
    <tr>
      <th>100019</th>
      <td>8.786688</td>
    </tr>
    <tr>
      <th>10002</th>
      <td>10.196485</td>
    </tr>
    <tr>
      <th>100020</th>
      <td>8.375616</td>
    </tr>
    <tr>
      <th>100021</th>
      <td>8.680761</td>
    </tr>
    <tr>
      <th>100022</th>
      <td>8.028204</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>10006</th>
      <td>10.341683</td>
    </tr>
    <tr>
      <th>100060</th>
      <td>10.098671</td>
    </tr>
    <tr>
      <th>100061</th>
      <td>10.384215</td>
    </tr>
    <tr>
      <th>100062</th>
      <td>9.506863</td>
    </tr>
    <tr>
      <th>100063</th>
      <td>10.610929</td>
    </tr>
    <tr>
      <th>100064</th>
      <td>9.799675</td>
    </tr>
    <tr>
      <th>100065</th>
      <td>9.793370</td>
    </tr>
    <tr>
      <th>100066</th>
      <td>10.207170</td>
    </tr>
    <tr>
      <th>100067</th>
      <td>9.205553</td>
    </tr>
    <tr>
      <th>100068</th>
      <td>9.259744</td>
    </tr>
    <tr>
      <th>100069</th>
      <td>9.528482</td>
    </tr>
    <tr>
      <th>10007</th>
      <td>11.197317</td>
    </tr>
    <tr>
      <th>100070</th>
      <td>8.990312</td>
    </tr>
    <tr>
      <th>100071</th>
      <td>9.478879</td>
    </tr>
    <tr>
      <th>100072</th>
      <td>9.778368</td>
    </tr>
    <tr>
      <th>100073</th>
      <td>14.046277</td>
    </tr>
    <tr>
      <th>100074</th>
      <td>9.069135</td>
    </tr>
    <tr>
      <th>100075</th>
      <td>8.941192</td>
    </tr>
    <tr>
      <th>100076</th>
      <td>8.190762</td>
    </tr>
    <tr>
      <th>100077</th>
      <td>9.983566</td>
    </tr>
    <tr>
      <th>100078</th>
      <td>10.470915</td>
    </tr>
    <tr>
      <th>100079</th>
      <td>9.880966</td>
    </tr>
    <tr>
      <th>10008</th>
      <td>10.126566</td>
    </tr>
    <tr>
      <th>100080</th>
      <td>9.670865</td>
    </tr>
    <tr>
      <th>100081</th>
      <td>10.406398</td>
    </tr>
    <tr>
      <th>100082</th>
      <td>10.390501</td>
    </tr>
    <tr>
      <th>100083</th>
      <td>9.173047</td>
    </tr>
    <tr>
      <th>100084<

In [78]:
asterank_dv_df["dv"].describe()

count    730078.000000
mean          9.918571
std           1.373073
min           3.826952
25%           9.037506
50%           9.883558
75%          10.583367
max          53.673297
Name: dv, dtype: float64

In [67]:
def get_asterank_pdes(row):
    
    splitted = row.split("(")
    if len(splitted)==1:
        return splitted[0]
    else:
        return splitted[1][:len(splitted[1])-1]  


jpl_df2["asterank_pdes"] = jpl_df2['full_name'].apply(lambda x : get_asterank_pdes(x))

In [68]:
jpl_df2[['pdes','full_name','asterank_pdes']][jpl_df2['full_name'].str.contains('\(')].head(3)

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pdes</th>
      <th>full_name</th>
      <th>asterank_pdes</th>
    </tr>
    <tr>
      <th>id</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>a0000330</th>
      <td>330</td>
      <td>330 Adalberta (A910 CB)</td>
      <td>A910 CB</td>
    </tr>
    <tr>
      <th>a0000333</th>
      <td>333</td>
      <td>333 Badenia (1892 A)</td>
      <td>1892 A</td>
    </tr>
    <tr>
      <th>a0000334</th>
      <td>334</td>
      <td>334 Chicago (1892 L)</td>
      <td>1892 L</td>
    </tr>
  </tbody>
</table>
</div>

In [94]:
jpl_df3 = pd.merge(jpl_df2,asterank_dv_df,how='left',left_on="pdes",right_index=True)

In [95]:
print("rows =%d"%len(jpl_df3))
print("columns = %d"%len(jpl_df3.columns))

rows =732354
columns = 68


In [96]:
jpl_df3["dv"].describe()

count    723329.000000
mean          9.902653
std           1.253921
min           3.826952
25%           9.035381
50%           9.879806
75%          10.578269
max          53.673297
Name: dv, dtype: float64

In [97]:
jpl_df3[['pdes','full_name','asterank_pdes','dv']].head(10)

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pdes</th>
      <th>full_name</th>
      <th>asterank_pdes</th>
      <th>dv</th>
    </tr>
    <tr>
      <th>id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>a0000001</th>
      <td>1</td>
      <td>1 Ceres</td>
      <td>1 Ceres</td>
      <td>10.533885</td>
    </tr>
    <tr>
      <th>a0000002</th>
      <td>2</td>
      <td>2 Pallas</td>
      <td>2 Pallas</td>
      <td>14.168266</td>
    </tr>
    <tr>
      <th>a0000003</th>
      <td>3</td>
      <td>3 Juno</td>
      <td>3 Juno</td>
      <td>9.683807</td>
    </tr>
    <tr>
      <th>a0000004</th>
      <td>4</td>
      <td>4 Vesta</td>
      <td>4 Vesta</td>
      <td>9.383595</td>
    </tr>
    <tr>
      <th>a0000005</th>
      <td>5</td>
      <td>5 Astraea</td>
      <td>5 Astraea</td>
      <td>9.206538</td>
    </tr>
    <tr>
      <th>a0000006</th>
      <td>6</td>
      <td>6 Hebe</td>
      <td>6 Hebe</td>
      <td>9.752704</td>
    </tr>
    <tr>
      <th>a0000007</th>
      <td>7</td>
      <td>7 Iris</td>
      <td>7 Iris</td>
      <td>8.598047</td>
    </tr>
    <tr>
      <th>a0000008</th>
      <td>8</td>
      <td>8 Flora</td>
      <td>8 Flora</td>
      <td>8.563821</td>
    </tr>
    <tr>
      <th>a0000009</th>
      <td>9</td>
      <td>9 Metis</td>
      <td>9 Metis</td>
      <td>9.172799</td>
    </tr>
    <tr>
      <th>a0000010</th>
      <td>10</td>
      <td>10 Hygiea</td>
      <td>10 Hygiea</td>
      <td>10.461966</td>
    </tr>
  </tbody>
</table>
</div>

In [101]:
len(jpl_df3[~jpl_df3["dv"].notnull()])

9025

These 9025 have a strange key mismatch. Instead of doing some magic to make things work, I will let them with Nan.

If later the feature prove to have value, them we can recover.

In [103]:
jpl_df3.to_csv("Data/Computed/jpl_population_20170408.csv")

In [4]:
jpl_df3 = pd.read_csv("Data/Computed/jpl_population_20170408.csv", low_memory=False, index_col='id')
jpl_df3.shape

(732354, 68)

In [5]:
jpl_df3.columns

Index(['spkid', 'full_name', 'pdes', 'name', 'neo', 'pha', 'H', 'G',
       'diameter', 'extent', 'albedo', 'rot_per', 'GM', 'BV', 'UB', 'IR',
       'spec_B', 'spec_T', 'H_sigma', 'diameter_sigma', 'orbit_id', 'epoch',
       'epoch_mjd', 'epoch_cal', 'equinox', 'e', 'a', 'q', 'i', 'om', 'w',
       'ma', 'ad', 'n', 'tp', 'tp_cal', 'per', 'per_y', 'moid', 'moid_ld',
       'moid_jup', 't_jup', 'sigma_e', 'sigma_a', 'sigma_q', 'sigma_i',
       'sigma_om', 'sigma_w', 'sigma_ma', 'sigma_ad', 'sigma_n', 'sigma_tp',
       'sigma_per', 'class', 'producer', 'data_arc', 'first_obs', 'last_obs',
       'n_obs_used', 'n_del_obs_used', 'n_dop_obs_used', 'condition_code',
       'rms', 'two_body', 'A1', 'A2', 'asterank_pdes', 'dv'],
      dtype='object')